features/ui/ui-kit-button/button.tsx

```diff
@@ -0,0 +1,90 @@
+import classNames from "classnames";
+import styles from "./button.module.scss";
+
+export type ButtonProps = {
+  children?: React.ReactNode;
+  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
+  as: "button" | "a";
+  href?: "string";
+  disabled?: boolean;
+  size: "small" | "medium" | "large" | "xlarge";
+  color:
+    | "primary"
+    | "secondary"
+    | "gray"
+    | "empty"
+    | "empty-gray"
+    | "error"
+    | "empty-error";
+  icon?: boolean;
+  iconSrc?: string;
+  iconAlt?: string;
+  iconPosition?: "leading" | "trailing" | "only";
+};
+
+export function UIButton({
+  children,
+  onClick,
+  as = "button",
+  href,
+  disabled = false,
+  size = "medium",
+  color = "primary",
+  icon = false,
+  iconSrc,
+  iconAlt,
+  iconPosition,
+}: ButtonProps) {
+  // const buttonClasses = `styles.button styles[${size}] styles[${color}] ${disabled ? 'styles.disabled' : ''}`;
+  // const iconOnly = iconPosition === 'only' ? styles["icon-only"] : ''
+  if (as === "button") {
```

Comment:
It's great that you identified the need for the `as` prop here. I forgot about that in the initial button task. Only added it in the landing page header task later.

However, it's not ideal to use this kind of if else statement as you basically have to duplicate everything only to render a different tag. There's a pattern called "polymorphic component" that allows you to take the `as` prop and use it as component. The trick is to assign the `as` prop to a const starting with a capital letter like `const Component = as`. The tricky part is TypeScript though. It took me some time and help of ChatGPT to get the types right 😄  Here's my code

```
import classNames from "classnames";
import styles from "./button.module.scss";

export enum ButtonSize {
  Small = "sm",
  Medium = "md",
  Large = "lg",
  XLarge = "xl",
}
export enum ButtonColor {
  Primary = "primary",
  Secondary = "secondary",
  Gray = "gray",
  Error = "error",
}
export enum ButtonVariant {
  Default = "default",
  Empty = "empty",
  IconOnly = "iconOnly",
}

const ButtonDefaultAsType = "button" as const;
type ButtonDefaultAsType = typeof ButtonDefaultAsType;

type ButtonOwnProps<As extends React.ElementType> = {
  size?: ButtonSize;
  color?: ButtonColor;
  variant?: ButtonVariant;
  as?: As;
};

type ButtonProps<As extends React.ElementType> = ButtonOwnProps<As> &
  Omit<React.ComponentProps<As>, keyof ButtonOwnProps<As>>;

export function Button<As extends React.ElementType = ButtonDefaultAsType>({
  className,
  size,
  color,
  variant,
  as,
  ...props
}: ButtonProps<As>) {
  const Component = as || ButtonDefaultAsType;
  return (
    <Component
      className={classNames(
        styles.button,
        size && styles[size],
        color && styles[color],
        variant && styles[variant],
        className,
      )}
      {...props}
    />
  );
}
```